### PR TITLE
[16.0][ADD] product_set

### PR DESCRIPTION
**Pull request by CRogos**
Splitting sale_product_set into product_set and sale_product_set as suggested by @pedrobaeza in https://github.com/OCA/sale-workflow/pull/2582

### CODE REVIEW

File: product_set/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+# Copyright 2018 Camptocamp, ACSONE SA/NV
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+{
+    "name": "Product set",
+    "category": "Sale",
+    "license": "AGPL-3",
+    "author": "Anybox, Odoo Community Association (OCA)",
+    "version": "16.0.1.0.0",
+    "website": "https://github.com/OCA/product-attribute",
+    "summary": "Sale product set",
```
Comment (pedrobaeza) [inline product_set/__manifest__.py]: Remove this line

File: product_set/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    "depends": ["product", "onchange_helper"],
```
Comment (pedrobaeza) [inline product_set/__manifest__.py]: I think `onchange_helper` is not needed here.

File: product_set/i18n/sale_product_set.pot
Comment (pedrobaeza) [inline product_set/i18n/sale_product_set.pot:1]: Remove this file

File: product_set/models/product_set_line.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
+# Copyright 2015 Anybox S.A.S
+# Copyright 2016-2018 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from odoo import fields, models
+
+
+class ProductSetLine(models.Model):
+    _name = "product.set.line"
```
Comment (pedrobaeza) [inline product_set/models/product_set_line.py:8]: This content can be inside `product_set.py` file, as it's a o2m model.

File: product_set/models/product_set_line.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    _description = "Product set line"
+    _rec_name = "product_id"
+    _order = "sequence"
```
Comment (pedrobaeza) [inline product_set/models/product_set_line.py]: For being more complete.

```suggestion
    _order = "product_set_id, sequence, product_id"
```

File: product_set/models/product_set_line.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        digits="Product Unit of Measure", required=True, default=1.0
+    )
+    product_set_id = fields.Many2one("product.set", string="Set", ondelete="cascade")
+    active = fields.Boolean(
+        string="Active", related="product_set_id.active", store=True, readonly=True
+    )
+    sequence = fields.Integer(required=True, default=0)
+    discount = fields.Float(string="Discount (%)", digits="Discount", default=0.0)
+
+    def prepare_sale_order_line_values(self, order, quantity, max_sequence=0):
```
Comment (pedrobaeza) [inline product_set/models/product_set_line.py]: This one should go in `sale_product_set`.

File: product_set/readme/DESCRIPTION.rst [hunk 6]
```diff
@@ -0,0 +1,11 @@
+A **product set** is a list of products which end customers aren't aware, this
+list is defined by sales manager.
+
+This module aims to help salesman to quickly create several sale order lines
```
Comment (pedrobaeza) [inline product_set/readme/DESCRIPTION.rst]: Adapt this description for being more general.

File: product_set/readme/USAGE.rst [hunk 7]
```diff
@@ -0,0 +1,19 @@
+To use this module, you need to:
```
Comment (pedrobaeza) [inline product_set/readme/USAGE.rst]: Something like this:

```suggestion
To use this module, you need to install subsequent modules like `sale_product_set` and check their instructions.
```

File: product_set/security/rule_product_set.xml [hunk 8]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <record id="product_set_comp_rule" model="ir.rule">
+        <field name="name">Product Set multi company rule</field>
+        <field name="model_id" ref="model_product_set" />
+        <field name="global" eval="True" />
+        <field
+            name="domain_force"
+        >['|',('company_id','=',False),('company_id','in',company_ids)]</field>
```
Comment (pedrobaeza) [inline product_set/security/rule_product_set.xml]: This is a bit more performant:

```suggestion
        >[('company_id', 'in', [False] + company_ids)]</field>
```

File: product_set/security/rule_product_set.xml [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+    </record>
+    <record id="product_set_line_comp_rule" model="ir.rule">
+        <field name="name">Product Set Line multi company rule</field>
+        <field name="model_id" ref="model_product_set_line" />
+        <field name="global" eval="True" />
+        <field
+            name="domain_force"
+        >['|',('product_set_id.company_id','=',False),('product_set_id.company_id','in',company_ids)]</field>
```
Comment (pedrobaeza) [inline product_set/security/rule_product_set.xml]: Should we put the field as related stored for performance reasons?

File: product_set/static/description/icon.svg
Comment (pedrobaeza) [inline product_set/static/description/icon.svg:1]: Remove this file, as it's the master template.

File: product_set/static/description/product_set.png
Comment (pedrobaeza) [inline product_set/static/description/product_set.png:1]: This screenshot should be in `sale_product_set`.

**Review by pedrobaeza — COMMENTED** [review]
Thanks for doing the split!

Please check my comments inline.

File: product_set/models/product_set_line.py (hunk 3 above)
Comment (CRogos) [reply product_set/models/product_set_line.py:8]: In my opinion leaving it in an separate file makes it easier when searching the repository.

Comment (pedrobaeza) [reply product_set/models/product_set_line.py:8]: Well, these models are always simpler, and very tied to the parent one. It's the usual distribution (see Odoo and OCA for this pattern), but I'm not forcing you to join it.

File: product_set/security/rule_product_set.xml (hunk 9 above)
Comment (CRogos) [reply product_set/security/rule_product_set.xml]: you mean this way?
company_id = fields.Many2one('res.company', related='product_set_id.company_id')

<field name="domain_force">[('company_id', 'in', [False] + company_ids)]</field>

Comment (pedrobaeza) [reply product_set/security/rule_product_set.xml]: Yes, but `store=True`

File: product_set/models/product_set_line.py (hunk 5 above)
Comment (pedrobaeza) [reply product_set/models/product_set_line.py]: What about this?

Comment (CRogos) [reply product_set/models/product_set_line.py]: you are right, did not see it. -> Done, but I've left the sequence in the product_set part.

File: product_set/readme/DESCRIPTION.rst (hunk 6 above)
Comment (CRogos) [reply product_set/readme/DESCRIPTION.rst]: Changed, but every improvement highly welcome.

**Comment by CRogos** [comment]
@rousseldenis tests fixed.

File: product_set/models/product_set_line.py (hunk 3 above)
Comment (lmignon) [reply product_set/models/product_set_line.py:8]: Personally I  prefer separate files :smirk:

**Comment by lmignon** [comment]
@CRogos No menu available into the UI ?

**Comment by pedrobaeza** [comment]
`product` is not providing menu as well. It's provided by sale/purchase, and there's no specific permissions yet.

**Review by lmignon — APPROVED** [review]
Code review only

File: product_set/readme/DESCRIPTION.rst [hunk 10]
```diff
@@ -0,0 +1,10 @@
+A **product set** is a list of products which are usually used together.
+
+This module aims to help adding several products quickly to sale order lines
```
Comment (pedrobaeza) [inline product_set/readme/DESCRIPTION.rst]:
```suggestion
This module aims to help defining several products under a name, for later
being added in a quick way into other document.
```

**Comment by simahawk** [comment]
@pedrobaeza good for you?

**Comment by lmignon** [comment]
Migration scripts are missing.... (here or in https://github.com/OCA/sale-workflow/pull/2604)

**Pull request merged**